src/plugins/kibana/public/dashboard/components/panel/panel.html

```diff
@@ -1,14 +1,34 @@
 <div class="panel panel-default" ng-switch on="panel.type" ng-if="savedObj || error">
   <div class="panel-heading">
-    <span class="panel-title" title="{{::savedObj.title}}">
+    <div class="panel-title" title="{{::savedObj.title}}">
       <i
         class="fa"
         ng-class="savedObj.vis.type.icon"
         aria-label="{{::savedObj.vis.type.title}} Icon"
         title="{{::savedObj.vis.type.title}}">
       </i>
-      {{::savedObj.title}}
-    </span>
+      <a class="action panel-title-edit" ng-hide="showFormTitle" ng-click="setTitle()">
+        <span ng-if="panel.title">{{panel.title}}</span>
+        <span ng-hide="panel.title">{{::savedObj.title}}</span>
+      </a>
+
+      <div>
+       <form class="inline-form" ng-if="showFormTitle" ng-show="chrome.getVisible()" role="form">
+          <input type="text" name="title" ng-model="panel.title" placeholder="{{::savedObj.title}}" input-focus>
```

Comment:
This needs the `form-control` class applied to it. You'll need to adjust other styles to keep this from wrapping.